File: tests/fixtures/python/repository.py
"""Data access layer with repository pattern."""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import List, Optional, Protocol
import threading
import sqlite3


class NotFoundError(Exception):
    """Raised when an entity is not found."""


class DuplicateError(Exception):
    """Raised when a duplicate entity is detected."""


class _ConnectionError(Exception):
    """Internal connection error."""


@dataclass
class Entity:
    """Base entity for all persisted objects."""
    id: str
    version: int = 0
    created_at: int = 0
    updated_at: int = 0


class Repository(ABC):
    """Abstract base for data access."""

    @abstractmethod
    def find_by_id(self, entity_id: str) -> Optional[Entity]:
        """Find an entity by ID."""

    @abstractmethod
    def find_all(self) -> List[Entity]:
        """Return all entities."""

    @abstractmethod
    def save(self, entity: Entity) -> None:
        """Persist an entity."""

    @abstractmethod
    def delete(self, entity_id: str) -> bool:
        """Delete an entity by ID."""


class Cacheable(Protocol):
    """Protocol for cacheable repositories."""

    def invalidate(self, key: str) -> None: ...
    def warm(self) -> None: ...


class SQLRepository(Repository):
    """SQL-backed repository with in-memory cache."""

    def __init__(self, db: sqlite3.Connection):
        self._db = db
        self._cache: dict = {}
        self._lock = threading.RLock()

    def find_by_id(self, entity_id: str) -> Optional[Entity]:
        """Find entity, checking cache first."""
        with self._lock:
            if entity_id in self._cache:
                return self._cache[entity_id]

        cursor = self._db.execute(
            "SELECT id, version FROM entities WHERE id = ?",
            (entity_id,)
        )
        row = cursor.fetchone()
        if row is None:
            return None

        entity = Entity(id=row[0], version=row[1])
        with self._lock:
            self._cache[entity_id] = entity
        return entity

    def find_all(self) -> List[Entity]:
        """Return all entities from database."""
        cursor = self._db.execute("SELECT id, version FROM entities")
        entities = []
        for row in cursor.fetchall():
            entities.append(Entity(id=row[0], version=row[1]))
        return entities

    def save(self, entity: Entity) -> None:
        """Save entity and update cache."""
        self._db.execute(
            "INSERT OR REPLACE INTO entities (id, version) VALUES (?, ?)",
            (entity.id, entity.version)
        )
        self._db.commit()
        with self._lock:
            self._cache[entity.id] = entity

    def delete(self, entity_id: str) -> bool:
        """Delete entity and invalidate cache."""
        cursor = self._db.execute(
            "DELETE FROM entities WHERE id = ?",
            (entity_id,)
        )
        self._db.commit()
        with self._lock:
            self._cache.pop(entity_id, None)
        return cursor.rowcount > 0

    def invalidate(self, key: str) -> None:
        """Remove a cache entry."""
        with self._lock:
            self._cache.pop(key, None)

    def warm(self) -> None:
        """Pre-load all entities into cache."""
        entities = self.find_all()
        with self._lock:
            for e in entities:
                self._cache[e.id] = e

    def __clear_cache(self):
        """Private: clear the entire cache."""
        self._cache.clear()


def create_repository(db_path: str) -> SQLRepository:
    """Factory to create a connected repository."""
    conn = sqlite3.connect(db_path)
    conn.execute(
        "CREATE TABLE IF NOT EXISTS entities (id TEXT PRIMARY KEY, version INTEGER)"
    )
    return SQLRepository(conn)


def _migrate_schema(conn: sqlite3.Connection) -> None:
    """Internal: run schema migrations."""
    conn.execute("PRAGMA journal_mode=WAL")

File: tests/fixtures/go/repository.go
// Package repository provides data access abstractions.
package repository

import (
	"context"
	"database/sql"
	"errors"
	"sync"
)

// Sentinel errors for repository operations.
var (
	ErrNotFound    = errors.New("entity not found")
	ErrDuplicate   = errors.New("duplicate entity")
	ErrConnection  = errors.New("connection failed")
	errPoolClosed  = errors.New("pool is closed")
)

// Entity is the base type for all persisted objects.
type Entity struct {
	ID        string
	Version   int
	CreatedAt int64
	UpdatedAt int64
}

// Repository defines CRUD operations.
type Repository interface {
	FindByID(ctx context.Context, id string) (*Entity, error)
	FindAll(ctx context.Context) ([]*Entity, error)
	Save(ctx context.Context, entity *Entity) error
	Delete(ctx context.Context, id string) error
}

// Cacheable adds caching contract.
type Cacheable interface {
	Invalidate(key string)
	Warm(ctx context.Context) error
}

// SQLRepository implements Repository with a SQL backend.
type SQLRepository struct {
	sync.RWMutex
	db     *sql.DB
	cache  map[string]*Entity
}

// NewSQLRepository creates a SQLRepository.
func NewSQLRepository(db *sql.DB) *SQLRepository {
	return &SQLRepository{
		db:    db,
		cache: make(map[string]*Entity),
	}
}

// FindByID looks up an entity by ID, checking cache first.
func (r *SQLRepository) FindByID(ctx context.Context, id string) (*Entity, error) {
	r.RLock()
	if cached, ok := r.cache[id]; ok {
		r.RUnlock()
		return cached, nil
	}
	r.RUnlock()

	row := r.db.QueryRowContext(ctx, "SELECT id, version FROM entities WHERE id = ?", id)
	entity := &Entity{}
	if err := row.Scan(&entity.ID, &entity.Version); err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, ErrNotFound
		}
		return nil, err
	}

	r.Lock()
	r.cache[id] = entity
	r.Unlock()

	return entity, nil
}

// FindAll returns all entities.
func (r *SQLRepository) FindAll(ctx context.Context) ([]*Entity, error) {
	rows, err := r.db.QueryContext(ctx, "SELECT id, version FROM entities")
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var entities []*Entity
	for rows.Next() {
		e := &Entity{}
		if err := rows.Scan(&e.ID, &e.Version); err != nil {
			return nil, err
		}
		entities = append(entities, e)
	}
	return entities, rows.Err()
}

// Save persists an entity and updates the cache.
func (r *SQLRepository) Save(ctx context.Context, entity *Entity) error {
	_, err := r.db.ExecContext(ctx, "INSERT INTO entities (id, version) VALUES (?, ?)", entity.ID, entity.Version)
	if err != nil {
		return err
	}

	r.Lock()
	r.cache[entity.ID] = entity
	r.Unlock()
	return nil
}

// Delete removes an entity and invalidates the cache.
func (r *SQLRepository) Delete(ctx context.Context, id string) error {
	_, err := r.db.ExecContext(ctx, "DELETE FROM entities WHERE id = ?", id)
	if err != nil {
		return err
	}

	r.Lock()
	delete(r.cache, id)
	r.Unlock()
	return nil
}

// Invalidate removes a cache entry.
func (r *SQLRepository) Invalidate(key string) {
	r.Lock()
	delete(r.cache, key)
	r.Unlock()
}

// Warm pre-loads all entities into cache.
func (r *SQLRepository) Warm(ctx context.Context) error {
	entities, err := r.FindAll(ctx)
	if err != nil {
		return err
	}

	r.Lock()
	for _, e := range entities {
		r.cache[e.ID] = e
	}
	r.Unlock()
	return nil
}

// newInternalHelper is unexported.
func newInternalHelper() string {
	return "helper"
}
